src/middleware/operation.rs

```diff
@@ -104,72 +132,95 @@ impl Operation {
             Self::TransitiveEqualFromStatements(_, _) => OT::Native(TransitiveEqualFromStatements),
             Self::GtToNotEqual(_) => OT::Native(GtToNotEqual),
             Self::LtToNotEqual(_) => OT::Native(LtToNotEqual),
-            Self::ContainsFromEntries(_, _) => OT::Native(ContainsFromEntries),
-            Self::NotContainsFromEntries(_, _) => OT::Native(NotContainsFromEntries),
+            Self::ContainsFromEntries(_, _, _, _) => OT::Native(ContainsFromEntries),
+            Self::NotContainsFromEntries(_, _, _) => OT::Native(NotContainsFromEntries),
             Self::SumOf(_, _, _) => OT::Native(SumOf),
             Self::ProductOf(_, _, _) => OT::Native(ProductOf),
             Self::MaxOf(_, _, _) => OT::Native(MaxOf),
             Self::Custom(cpr, _) => OT::Custom(cpr.clone()),
         }
     }
 
-    pub fn args(&self) -> Vec<Statement> {
+    pub fn args(&self) -> Vec<OperationArg> {
```

Comment:
My suggestion would be to change this so that `args` still returns `Vec<Statement>`.
And instead of introducing an `OperationArg`, add a new method `Operation::aux(&self) -> OperationAux` which returns the auxiliary data of the operation.

```rust
enum OperationAux {
  None,
  MerkleProof(MerkleProof),
}
```

I think this would simplify things a bit.

Then on the frontend I would suggest doing something similar.

```rust
pub struct Operation(pub OperationType, pub Vec<OperationArg>, pub OperationAux);
```
And keep the `frontend::OperationArg` untouched.

And the same idea in mock_main